### ISSUE TITLE
Random Search problems

### ISSUE THREAD

**Issue by mboussaa** [MOEAFramework/MOEAFramework#26]
NondominatedPopulation result = new Executor()
	    .withProblemClass(Evaluation.class)
	    .withAlgorithm("Random")
	    .withMaxEvaluations(20)
	    .run();

This runs the program with Random Search for more than 20 iterations. Why max evaluations property is not applied?

if I want to evaluate each solution generated by Random search according to two objectives how is it possible? 

Thanks

**Comment by dhadka** [comment]
It also uses the "populationSize" argument, which is set to 100 by default.  The number of generated solutions will therefore be a multiple of 100.  (This is done for parallelization so you can control how many solutions are generated and evaluated in parallel).  If you only want to generate 20, use:

~~~~~~
		NondominatedPopulation result = new Executor()
				.withProblem("UF1")
				.withAlgorithm("Random")
				.withMaxEvaluations(20)
				.withProperty("populationSize", 20)
				.run();
~~~~~~

I'll look into changing this behavior since it is confusing and not what one would expect.  Thanks for pointing this out.

Random search generates N uniformly-distributed solutions, calls your `Evaluation` class' `evaluate(Solution)` method on each solution to compute the objectives, and adds the solutions to a Pareto non-dominated set.  This Pareto non-dominated set is returned as the result.  For problems with two or more objectives, there will typically be more than one optimal solution.

**Comment by mboussaa** [comment]
Hi,
Thanks for these tips. It works. However, it is better to have more details in the documentation (manual). 
So you said that if there are two or more objectives it will return the set of pareto optimal solutions. RS does that according to which indicator . Is it hypervolume like in NSGAII?

**Comment by dhadka** [comment]
It uses the Pareto dominance relation (no indicator is used).  Essentially, if you run random search for 1000 evaluations, it will generate 1000 random solutions.  Then, it finds the Pareto frontier --- the solutions that are not Pareto dominated by any other solution.  This Pareto frontier (sometimes also called Pareto non-dominated solutions or Pareto optimal solutions) is the result.

See https://en.wikipedia.org/wiki/Pareto_efficiency#Use_in_engineering for more details.

**Comment by mboussaa** [comment]
+1. So I guess default NSGAII has no indicator as well... Right? The indicators are only useful for algorithms comparison ?

**Comment by dhadka** [comment]
Yes, NSGA-II uses fast non-dominated sorting (Pareto dominance with crowding distance).  Some of the algorithms do use indicators, such as IBEA, HypE, and SMS-EMOA.

**Comment by mboussaa** [comment]
Thank you. It looks pretty clear now,
In fact, I have question and I don't know if I have to open a new issue (it is not really an issue). I have my own heuristic that looks like random search. I implemented it manually without the use of MOEA. This heuristic generate random solutions let's say. Each solution has two objectives. Then, I would like to compare pareto front solution got using NSGAII and RS (from MOEA) to my own generated solutions. In other words, I want to compare/evaluate each solution I have generated manually to pareto front solutions generated by MOEA(RS + NSGA). How is it possible ? 

Which part of MOEA Framework I can use to evaluate external solutions ?

Thanks in advance.

**Comment by dhadka** [comment]
As an example, first you would need to load your external solutions.  Lets say you have a text file containing the decision variables for each of your solutions (one solution per line):

```
0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0
1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0
```

You can load this data with:

```java

		Problem problem = new UF1(); // put your problem here
		File file = new File("solutions.txt");
		
		// load your solutions from file
		String line = null;
		BufferedReader reader = null;
		Population manualSolutions = new Population();
		
		try {
			reader = new BufferedReader(new FileReader(file));
			
			while ((line = reader.readLine()) != null) {
				Solution solution = problem.newSolution();
				String[] tokens = line.split("\\s+"); // split by whitespace
				
				for (int i = 0; i < problem.getNumberOfVariables(); i++) {
					EncodingUtils.setReal(solution.getVariable(i), Double.parseDouble(tokens[i]));
				}
				
				manualSolutions.add(solution);
			}
		} finally {
			if (reader != null) {
				reader.close();
			}
		}
		
		// evaluate the solutions
		for (Solution solution : manualSolutions) {
			problem.evaluate(solution);
		}
```

Also lets say you have the solutions from NSGA-II:

```java

		NondominatedPopulation nsgaResults = new Executor()
				.withProblemClass(UF1.class) // put your problem here
				.withAlgorithm("NSGAII")
				.withMaxEvaluations(10000)
				.run();
```

Then plot the results (requires the latest master branch on Github):

```java

		new Plot()
				.add("External", manualSolutions)
				.add("NSGA-II", nsgaResults)
				.show();
```

If you want to see how many solutions from each set are non-dominated:

```java

		ReferenceSetMerger merger = new ReferenceSetMerger();
		merger.add("Manual", manualSolutions);
		merger.add("NSGA-II", nsgaResults);
		
		System.out.println("External: " + merger.getContributionFrom("External") + " / " + manualSolutions.size());
		System.out.println("NSGA-II: " + merger.getContributionFrom("NSGA-II") + " / " + nsgaResults.size());
```

There are likely several other ways to compare, but this should be enough to get your started.

**Comment by mboussaa** [comment]
Hi, Thank you for this feedback. However, I didn't get how to plot the pareto frontier using your method.

**Comment by dhadka** [comment]
The `ReferenceSetMerger` class takes two or more sets of solutions and combines them to produce a single Pareto front.  It also records where each solution in the final Pareto front originated.  In the following, we add two sets.  One would be your manually generated solutions and the other is the results produced by NSGA-II:

```java

        ReferenceSetMerger merger = new ReferenceSetMerger();
        merger.add("Manual", manualSolutions);
        merger.add("NSGA-II", nsgaResults);
```

Then we can see how many of your manually generated solutions are in the combined Pareto front (i.e., non-dominated with respect to NSGA-II's results):

```java

    System.out.println(merger.getContributionFrom("Manual").size());
```

You can also print the individual solutions:

```java

    for (Solution solution : merger.getContributionFrom("Manual")) {
        // print the decision variables and/or objectives
    }
```

We use the names "Manual" and "NSGA-II" to label which solutions originated from each set.  Calling `getContributionFrom("Manual")` returns the subset of solutions from the "Manual" set that remain in the combined Pareto front.

**Comment by mboussaa** [comment]
if I run my manual algorithm 10 times and gather pareto front solutions each time. How can I compare it statistically to RS and NSGAII using the analyzer?
like this example :
``
String problem = "UF1";
String[] algorithms = { "NSGAII", "GDE3", "eMOEA" };
         
//setup the experiment
Executor executor = new Executor()
    .withProblem(problem)
    .withMaxEvaluations(10000);
         
Analyzer analyzer = new Analyzer()
    .withProblem(problem)
    .includeHypervolume()
    .showStatisticalSignificance();
 
//run each algorithm for 50 seeds
for (String algorithm : algorithms) {
    analyzer.addAll(algorithm, 
        executor.withAlgorithm(algorithm).runSeeds(50));
}
 
//print the results
analyzer.printAnalysis();

**Comment by dhadka** [comment]
Something like the following should work, where `loadManualSolutions(i)` uses the code from the previous answer to load the solutions.

```java
Analyzer analyzer = new Analyzer()
    .withProblem(problem)
    .includeHypervolume()
    .showStatisticalSignificance();

for (String algorithm : algorithms) {
    analyzer.addAll(algorithm, 
        executor.withAlgorithm(algorithm).runSeeds(50));
}

for (int i = 0; i < 50; i++) {
    Population manualSolutions = loadManualSolutions(i);
    analyzer.add("Manual", new NondominatedPopulation(manualSolutions));
}

analyzer.printAnalysis();
```

**Comment by mboussaa** [comment]
Thank you @dhadka.

I have a question about the hypervolume indicator for algo comparison.

I am running my programs for only one time (seed = 1) 

Is it possible to get HV metric for only one run So that I can compare it with other algorithms?

**Comment by dhadka** [comment]
Certainly, you should be able to use the `Analyzer` (it'll just print out the min/median/max as the same value).  You can also look into using the [Hypervolume](http://moeaframework.org/javadoc/org/moeaframework/core/indicator/Hypervolume.html) class directly, but using the `Analyzer` is probably easier.

The coordinate of the HV reference point is computed as `max + delta*(max-min)`, where `delta` is 0.01 by default and min/max are the minimum and maximum values for each objective.

**Comment by mboussaa** [comment]
![capture d ecran 2016-03-01 a 12 03 14](https://cloud.githubusercontent.com/assets/5878565/13430633/269afef8-dfc6-11e5-8923-55b6672a5507.png)
To follow up with very nice discussion, I got the chart of pareto solutions (green and yellow). In red, the manual solutions.
I would like to plot the pareto frontier composing those points in green and yellow. Is it possible ? And then add labels to some manual points

For HV metric yes it works fine for one single seed. Thanks

**Comment by dhadka** [comment]
Q1: I'm not sure exactly what you mean by plotting approximations of the Pareto frontier.  Do you want to draw a curve estimating the Pareto frontier?  [Here is an example of adding line plots](http://keyboardscientist.weebly.com/blog/generating-simple-plots).  Labels/annotations are currently not supported, however since the [Plot](http://moeaframework.org/javadoc/org/moeaframework/analysis/plot/Plot.html) class uses JFreeChart, it's possible with some code modifications to add [text annotations](http://www.jfree.org/jfreechart/api/javadoc/org/jfree/chart/annotations/package-summary.html).

Q2: Certainly.  The first step would be to collect the Pareto frontier at the desired NFE.  Use the instrumenter to collect the approximation sets at a fixed interval:

```java

Instrumenter instrumenter = new Instrumenter()
     .withProblem(problemName)
     .attachApproximationSetCollector()
     .withFrequency(100);
```

Then read the values at the desired NFE:

```java

Accumulator accumulator = instrumenter.getLastAccumulator();

Population pfAt100NFE = new Population((Iterable<Solution>)accumulator.get("Approximation Set", 0));
Population pfAt500NFE = new Population((Iterable<Solution>)accumulator.get("Approximation Set", 4));
Population pfAt1000NFE = new Population((Iterable<Solution>)accumulator.get("Approximation Set", 9));
```

See [Example 3](http://moeaframework.org/examples.html#example3) for more details on using the Instrumenter.

**Comment by mboussaa** [comment]
![hiperbole](https://cloud.githubusercontent.com/assets/5878565/13462468/dcd72f90-e087-11e5-81d8-a5858dead72d.png)
yes something like the blue line in this figure that represents the curve of pareto front solutions

**Comment by dhadka** [comment]
It might be easier to create the plot in another tool, such as Excel, that already provides curve fitting functionality.

If not, you would first need to find an equation for the curve using some tool, such as [http://mycurvefit.com/](http://mycurvefit.com/), then generate the line data to plot:

```java

		int length = 100;
		double[] x = new double[length];
		double[] y = new double[length];
		double minX = 0.0;
		double maxX = 5.0;
		
		for (int i = 0; i < length; i++) {
			x[i] = (maxX - minX)*(i / (length-1.0)) + minX;
			y[i] = 7.24 + 0.35*x[i] + 0.014*x[i]*x[i] + 0.91*x[i]*x[i]*x[i];
		}
		
		new Plot().line("PF Fit",x, y).show();
```

**Comment by mboussaa** [comment]
CurveFit is quite awesome. Thank You David. I have another question. I don't know if I should open a new issue or not. I can edit this comment and open a new issue if you want.
Any way, my question is the following:

I am generating manual solutions (using an algorithm I've created by myself) and evaluating them using:
```
 for (Solution solution : manualSolutions) {
            problem.evaluate(solution);
        }
```
Then, I run NSGA-II and Random as known in MOEA Framework.
Afterthat, I use the merger to check if there are non dominated solution among those I ve created manually with comparison with RS and NSGAII
```
ReferenceSetMerger merger = new ReferenceSetMerger();
        merger.add("Manual", manualSolutions);
        merger.add("NSGA-II", nsgaResults);
```

If there is at least one manual solution that is non dominated I catch it:
```
 for (Solution solution : merger.getContributionFrom("Manual")) {
        // print the decision variables and/or objectives
    }
```
If not I can catch best solution  among those manual solution (In case `merger.getContributionFrom("Manual")` returns 0 entry) right? 

So my question is how can I catch the best solution among best manual ones and that even in case this solution worst than RS and NSGAII?

**Comment by dhadka** [comment]
You could put your manual solutions into a [NondominatedPopulation](http://moeaframework.org/javadoc/org/moeaframework/core/NondominatedPopulation.html) to get the Pareto frontier of only those solutions.

**Comment by emelmdn** [comment]
I am trying to see the results of Knapsack problem solved by 3 algorithms on the diagnostic tool. I tried to insert the Knapsack problem into the src package as described in the following link but I think something is not right since I get "no Provider for Knapsack" once I hit the run button on the diagnostic tool. I copied the Knapsack.java file (from the examples) into the src package into the org.moeaframework.problem.knapsack folder. Many thanks in regards...

**Comment by JavierRoman** [comment]
Hello. 

Is it possible to plot the pareto-optimal front of a problem? 

I'm trying to evaluate PAES on Zitzler problems and i'd like to have the global pareto-optimal front of each problem with the solutions got by PAES in the same plot. 

Thanks.
